kzg/src/eth/eip_7594.rs

```diff
@@ -0,0 +1,146 @@
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+use alloc::{format, string::String, vec::Vec};
+
+use crate::{
+    cfg_iter,
+    das::{EcBackend, DAS},
+    eip_4844::bytes_to_blob,
+    eth::{
+        BYTES_PER_BLOB, BYTES_PER_CELL, BYTES_PER_COMMITMENT, BYTES_PER_FIELD_ELEMENT,
+        BYTES_PER_PROOF, CELLS_PER_EXT_BLOB, FIELD_ELEMENTS_PER_CELL, FIELD_ELEMENTS_PER_EXT_BLOB,
+    },
+    Fr, G1,
+};
+
+pub type CellsKzgProofs = (Vec<[u8; BYTES_PER_CELL]>, Vec<[u8; BYTES_PER_PROOF]>);
+
+pub fn recover_cells_and_kzg_proofs_raw<B: EcBackend>(
+    cell_indices: &[usize],
+    cells: &[[u8; BYTES_PER_CELL]],
+    das: &impl DAS<B>,
+) -> Result<CellsKzgProofs, String>
+where
+    B::G1: Copy,
+    B::Fr: Copy,
+{
+    let nested_cells = cfg_iter!(cells)
+        .map(|bytes| bytes_to_cell::<B>(bytes))
+        .collect::<Result<Vec<Vec<_>>, _>>()?;
+    let cells = nested_cells.into_iter().flatten().collect::<Vec<_>>();
```

Comment:
I'm not sure, but looks like it would be better to do `flatten` first, and then conversion - that way you can avoid heap allocation (collecting each cell to vector), and also flattening will be zero-cost, as memory layout is not changed:
```suggestion
    let cells = cfg_iter!(cells.flatten()).map(B::Fr::from_bytes).collect::<Result<Vec<_>, String>>()?;
```

Reply:
flatten is not available in the slice array, which should be the better way?

Reply:
but, even if we do so, the bytes size will not fit into the field element, we need to chunk it first anyway

Reply:
Yeah, you're right, too rushed with this suggestion. It should be:
```suggestion
    let cells = cfg_iter!(cells.as_flattened().chunks(BYTES_PER_FIELD_ELEMENT)).map(B::Fr::from_bytes).collect::<Result<Vec<_>, _>>()?;
```
The point was that we can avoid unnecessary heap allocations, and keep only one

Reply:
I could not use `cfg_iter` on chunks, so remove on `cells` for now, will improve it later

Reply:
Ah right, chunks is already an iterator. There is [`par_chunks`](https://docs.rs/rayon/latest/rayon/slice/trait.ParallelSlice.html#method.par_chunks) in rayon, but I believe it wouldn't give much performance improvement